xmodule/services.py

```
@@ -308,3 +312,147 @@ def _handle_deprecated_progress_event(self, block, event):
         # in order to avoid duplicate work and possibly conflicting semantics.
         if not getattr(block, 'has_custom_completion', False):
             self.completion_service.submit_completion(block.scope_ids.usage_id, 1.0)
+
+
+class ProblemFeedbackService(Service):
+    """
+    An XBlock Service that allows XModules to define:
+
+    * If the answers are visible.
+    * Whether correctness is currently hidden. When correctness is hidden, this
+      limits the user's access to the correct/incorrect flags, messages, problem
+      scores, and aggregate subsection and course grades.
+    """
+    def __init__(self, block=None, user_is_staff=False, **kwargs):
+        super().__init__(**kwargs)
+        self._user_is_staff = user_is_staff
+        # This is needed because the `Service` class initialization expects the XBlock passed as an `xblock` keyword
+        #  argument, but the `service` method from the `DescriptorSystem` passes a `block`.
+        self._xblock = self.xblock() or block
+
+    def is_past_due(self):
+        """
+        Returns if the problem is past its due date.
+
+        This method returns True when an XBlock doesn't have a due date. That
+        is because if an XBlock doesn't have a due date, and we are trying to
+        assess whether we can display a correct answer to a learner, we want to
+        treat an answered problem without a due date as past due. Otherwise, we
+        can never show the answer to a learner, since there is no due date.
+
+        There are scenarios where you would want is_past_due to return False
+        when an XBlock doesn't have a due date, for example, when determining
+        whether a learner should still be able to submit their answer.
+        """
+        if not self._xblock:
+            return False
```

While raising an exception here would cause a runtime error affecting the learners, this check can result in a silent failure that's hard to detect. I wonder when this scenario (i.e., missing XBlock) can happen. Should we log an error in such cases?